src/psyclone/psyir/transformations/extract_trans.py

```diff
@@ -194,29 +207,15 @@ def apply(self, nodes, options=None):
         # the first child to be enclosed as that will be the position
         # of the ExtractNode.
         node_parent = node_list[0].parent
-        node_position = node_list[0].position
 
         # Create a Memento of the Schedule and the proposed
         # transformation
         schedule = node_list[0].root
-
         keep = Memento(schedule, self)
 
-        from psyclone.extractor import ExtractNode
-        extract_node = ExtractNode(parent=node_parent, children=node_list[:])
-
-        # Change all of the affected children so that they have the
-        # ExtractNode as their parent. Use a slice of the list of Nodes
-        # so that we're looping over a local copy of the list. Otherwise
-        # things get confused when we remove children from the list.
-        for child in node_list[:]:
-            # Remove child from the parent's list of children
-            node_parent.children.remove(child)
-            child.parent = extract_node
-
-        # Add the ExtractNode as a child of the parent of the Nodes being
-        # enclosed at the original location of the first of these Nodes
-        node_parent.addchild(extract_node,
-                             index=node_position)
+        # Pass the options to the constructor, used e.g. for the
+        # 'create_driver' flag.
+        self._node_class(parent=node_parent, children=node_list[:],
```

Comment:
Please add a comment e.g. "Call the constructor of the class of Node supplied to the constructor of this transformation. This creates a Node of .... with children..." I think we need this as the fact that the constructor modifies both node_parent and the children objects is non-intuitive (although it might have been me who introduced such functionality?).

Reply:
I can't remember who introduced that, but there were several instances of code doing the same thing, so I moved some of it in a base class. And tbh that should actually be further improved (i.e. it should go to Node), but we have #645 (and likely #294) for that. I've documented what is happening in more detail.